fix(containers): fluid containers

jdrueckert opened the pull request:
- reintroduce outcommented code in `FluidAuthoritySystem`
- use liquid detection methods from `Block.java` instead of `WorldProvider`

Helps fixing https://github.com/Terasology/JoshariasSurvival/issues/27 (at least the water part)

jdrueckert committed:
fix(containers): fluid containers

- reintroduce outcommented code in `FluidAuthoritySystem`
- use liquid detection methods from `Block.java` instead of `WorldProvider`

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -26,15 +30,18 @@
+@RegisterSystem(RegisterMode.AUTHORITY)
 public class FluidAuthoritySystem extends BaseComponentSystem {
+
+    private static final Logger logger = LoggerFactory.getLogger(FluidAuthoritySystem.class);
```
```suggestion
    private static final Logger logger = LoggerFactory.getLogger(FluidAuthoritySystem.class);
    
```

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -54,31 +61,30 @@
+        if (fluidContainer.fluidType == null || fluidContainer.volume < fluidContainer.maxVolume) {
+            Vector3f location = event.getInstigatorLocation();
+            Vector3f direction = new Vector3f(event.getDirection());
+            direction.normalize();
```
Could these two lines be merged? Or is it only JOML that allows for it?

```suggestion
            Vector3f direction = new Vector3f(event.getDirection()).normalize();
```

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -54,31 +61,30 @@
+            Vector3f location = event.getInstigatorLocation();
+            Vector3f direction = new Vector3f(event.getDirection());
+            direction.normalize();
+            for (int i = 0; i < 3; i++) {
+                location.add(direction);
+
+                Block block = worldProvider.getBlock(new Vector3i(location, 0.5f));
+                if (block.isLiquid() ) {
```
Could we refactor this loop to be contained in a method with a comprehensive name?

```java
private Optional<Block> getLiquidInReach(final Vector3f start, final Vector3f direction, int distance) {
	Vector3f location = new Vector3f(start);
	for (int i = 0; i < distance; i++) {
		location.add(direction);
        Block block = worldProvider.getBlock(new Vector3i(location, 0.5f));
		if (block.isLiquid()) {
			return Optional.of(block);
		}
	}
	return Optional.empty();
}
```

There may also be a check missing if there is water in reach, but hidden behind a solid block... that could easily be integrated to this new method.

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -54,31 +61,30 @@
+                if (block.isLiquid() ) {
+                    EntityRef owner = item.getOwner();
+                    final EntityRef removedItem = inventoryManager.removeItem(owner, event.getInstigator(), item, false, 1);
+                    if (removedItem != null && block.isWater()) {
```
```suggestion
					//TODO: replace with better fluid handling, maybe by new CoreFluids module
                    if (removedItem != null && block.isWater()) {
```

jdrueckert committed:
chore: integrate feedback

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -42,13 +44,29 @@
     @In
     private InventoryManager inventoryManager;
 
+    private Optional<Block> getLiquidInReach(final Vector3f start, final Vector3f direction, int distance) {
```
Please add a bit of Javadoc, as this method has a contract that's not obious at first sight (the direction is normalized).

```suggestion
	/**
	 * Search for a reachable liquid block in the given direction.
	 * 
	 * A liquid block is reachable iff it is within the specified distance from the starting location and 
	 * there is no penetrable block preventing direct access.
	 * If the returned option is non-empty, the block is guaranteed to be a liquid block.
	 *
	 * @param start		the starting location
	 * @param direction	the normalized direction vector
	 * @param distance	the reachable distance in number of blocks
	 *
	 * @return option of the liquid block found in reach, empty if none was found
	 */
    private Optional<Block> getLiquidInReach(final Vector3f start, final Vector3f direction, int distance) {
```

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -63,27 +81,24 @@ public void fillFluidContainerItem(ActivateEvent event, EntityRef item, FluidCon
-                        // Set the contents of this fluid container and fill it up to max capacity.
-                        FluidUtils.setFluidForContainerItem(removedItem, "Fluid:Water",
-                                removedItem.getComponent(FluidContainerItemComponent.class).maxVolume);
+            Optional<Block> block = getLiquidInReach(location, direction, 3);
+            if (block.isPresent() && block.get().isLiquid()) {
```
The `isLiquid()` property is guaranteed by `getLiquidInReach`, thus this check can be simplified.

```suggestion
            Optional<Block> block = getLiquidInReach(location, direction, 3);
            if (block.isPresent()) {
```

skaldarnar replied:
Instead of using the pattern 

```java
Optional<Block> block = ....;
if (block.isPresent()) {
	... block.get()....
}
```

you could use [`Optional::ifPresent`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#ifPresent-java.util.function.Consumer-) here:

```suggestion
            getLiquidInReach(location, direction, 3).ifPresent(block -> {
```

skaldarnar commented on src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:
```diff
@@ -63,27 +81,24 @@ public void fillFluidContainerItem(ActivateEvent event, EntityRef item, FluidCon
+                EntityRef owner = item.getOwner();
+                final EntityRef removedItem = inventoryManager.removeItem(owner, event.getInstigator(), item, false, 1);
+                //TODO: replace with better fluid handling, maybe by new CoreFluids module
+                if (removedItem != null && block.get().isWater()) {
```
When using `Optional::ifPresent` this becomes

```suggestion
                if (removedItem != null && block.isWater()) {
```

jdrueckert committed:
chore: integrate feedback

- add java docs
- move vector normalization into `getLiquidInReach` function
- improve `Optional` usage pattern
